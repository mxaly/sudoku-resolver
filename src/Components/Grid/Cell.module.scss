@import "@/styles/functions";

.cell {
  --text-color: var(--c-text-primary);

  display: flex;
  align-items: center;
  justify-content: center;

  aspect-ratio: 1/1;

  font-size: 4vmin;
  color: var(--text-color);

  background-color: var(--c-bg-elevated);
  border-radius: 3px;

  &.isConstant {
    --text-color: var(--c-white);
  }
}

.optionsGrid {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-template-rows: repeat(3, 1fr);

  width: 100%;
  height: 100%;

  color: var(--c-text-placeholder);
}

.option {
  display: flex;
  align-items: center;
  justify-content: center;

  aspect-ratio: 1/1;

  font-size: 1.9vmin;
}
