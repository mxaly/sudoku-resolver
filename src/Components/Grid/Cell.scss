.cell {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 1px;
  border-radius: var(--br-s);
  background-color: var(--c-primary-bg);
  font-size: var(--fs-50);
  color: var(--c-text-primary);

  // 3rd and 6th column/row have margin to separate 3x3 boxes
  &:nth-child(9n + 3),
  &:nth-child(9n + 6) {
    margin-right: 4px;
  }
  &:nth-child(n + 19):nth-child(-n + 27),
  &:nth-child(n + 46):nth-child(-n + 54) {
    margin-bottom: 4px;
  }

  &.is-constant {
    color: var(--c-text-secondary-dark);
  }
}

.options-grid {
  width: 100%;
  height: 100%;
  display: grid;

  grid-template-rows: repeat(3, 1fr);
  grid-template-columns: repeat(3, 1fr);

  gap: 0px;

  color: var(--c-text-secondary-dark);
}

.option {
  display: flex;
  justify-content: center;
  align-items: center;

  font-size: var(--fs-10);
  line-height: var(--fs-10);
}
