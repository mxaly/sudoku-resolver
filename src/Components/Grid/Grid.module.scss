.gridContainer {
  aspect-ratio: 1/1;
  width: 100%;
  max-width: 660px;
  max-height: 660px;
}

.grid {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-template-rows: repeat(3, 1fr);
  gap: 5px;

  width: 100%;
  height: 100%;
}
