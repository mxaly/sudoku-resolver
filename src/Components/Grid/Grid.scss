.grid-container {
  max-width: 660px;
  width: 100%;
  aspect-ratio: 1/1 !important;
}

.grid {
  height: 100%;
  width: 100%;

  display: grid;
  grid-template-rows: repeat(9, 1fr);
  grid-template-columns: repeat(9, 1fr);
  gap: 2px;
  padding: 2px;
}
