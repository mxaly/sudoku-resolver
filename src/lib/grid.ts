export type sudokuValue = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

export type cellAddress = [number, number];

interface CellData {
  value: sudokuValue | null;
  options: sudokuValue[];
  isConstant: boolean;
}

export interface Cell extends CellData {
  row: number;
  col: number;
  box: number;
}

export type Grid = Cell[];

const GRID_SIZE = 9;
export const OPTIONS = [...Array(9).keys()].map((k) => (k + 1) as sudokuValue);

export function createGrid(): Cell[] {
  const cells: Cell[] = [];
  for (let row = 0; row < GRID_SIZE; row++) {
    for (let col = 0; col < GRID_SIZE; col++) {
      cells.push(createCell({ row, col }));
    }
  }
  return cells;
}

export function createFromString(data: string): Cell[] {
  const grid = createGrid();
  data
    .replaceAll(/\s/g, "")
    .split("")
    .forEach((val, index) => {
      const row = Math.floor(index / 9);
      const col = index - row * 9;
      const num = parseInt(val);

      if (num > 0) {
        set(grid, [row, col], {
          value: num as sudokuValue,
          isConstant: true,
        });
      }
    });
  return grid;
}

function createCell({
  row,
  col,
  value,
  options,
  isConstant,
}: {
  row: number;
  col: number;
  value?: sudokuValue | null;
  options?: sudokuValue[];
  isConstant?: boolean;
}): Cell {
  return {
    value: value || null,
    options: options || [],
    row,
    col,
    box: getBoxIndex([row, col]),
    isConstant: !!isConstant,
  };
}

export function get(grid: Grid, addr: cellAddress): Cell {
  return grid[getIndex(addr)];
}

export function set(grid: Grid, addr: cellAddress, data: Partial<CellData>) {
  const cell = grid[getIndex(addr)];
  grid[getIndex(addr)] = { ...cell, ...data };
  return [...grid];
}

function getBoxIndex([rowIndex, colIndex]: cellAddress): number {
  const row = Math.floor(rowIndex / 3);
  const column = Math.floor(colIndex / 3);
  return row * 3 + column;
}

function getIndex([row, col]: [number, number]) {
  return row * GRID_SIZE + col;
}
