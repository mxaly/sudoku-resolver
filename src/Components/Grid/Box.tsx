import { Cell as CellData } from "@/lib/grid";
import { Cell } from "./Cell";

import styles from "./Box.module.scss";

interface Props {
  cells: CellData[];
}

export function Box({ cells }: Props) {
  const renderCells = cells.map(({ row, col, value, options, isConstant }) => {
    return (
      <Cell
        key={`${row}${col}`}
        value={value}
        options={options}
        isConstant={isConstant}
      />
    );
  });

  return <div className={styles.box}>{renderCells}</div>;
}
