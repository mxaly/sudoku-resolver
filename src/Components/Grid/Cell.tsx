import { sudokuValue } from "@/lib/grid";
import classNames from "classnames";

import style from "./Cell.module.scss";

const EMPTY_OPTIONS = Array(9).fill(null);

interface Props {
  value: sudokuValue | null;
  options: sudokuValue[];
  isConstant: boolean;
}
export function Cell({ value, options, isConstant }: Props) {
  const optionsList = EMPTY_OPTIONS.map((opt, index) => (
    <div className={style.option} key={index}>
      {options.includes((index + 1) as sudokuValue) ? index + 1 : ""}
    </div>
  ));

  const cellClass = classNames({
    [style.cell]: true,
    [style.isConstant]: isConstant,
  });

  return (
    <div className={cellClass} data-testid="cell">
      {value ? value : <div className={style.optionsGrid}>{optionsList}</div>}
    </div>
  );
}
