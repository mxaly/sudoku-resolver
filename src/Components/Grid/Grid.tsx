import { Cell as CellData } from "@/lib/grid";
import { Cell } from "./Cell";

import "./Grid.scss";

interface Props {
  cells: CellData[];
}

export function Grid({ cells }: Props) {
  const renderCells = cells.map((c) => (
    <Cell
      key={`${c.row}${c.col}`}
      value={c.value}
      options={c.options}
      isRightEdge={isRightEdgeCell(c)}
      isBottomEdge={isBottomEdgeCell(c)}
      isConstant={c.isConstant}
    />
  ));

  return (
    <div className="grid-container">
      <div className="grid">{renderCells}</div>
    </div>
  );
}

function isRightEdgeCell(cell: CellData) {
  return (cell.col + 1) % 3 === 0;
}

function isBottomEdgeCell(cell: CellData) {
  return (cell.row + 1) % 3 === 0;
}
