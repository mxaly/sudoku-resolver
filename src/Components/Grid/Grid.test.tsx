import { createFromString } from "@/lib/grid";
import { render } from "@testing-library/react";
import { Grid } from "./Grid";

describe("Grid", () => {
  it("should render the grid", () => {
    const grid = createFromString(`
    123456789 
    234567891 
    345678912

    456789123
    567891234
    678912345

    789123456
    891234567
    912345678`);

    const wrapper = render(<Grid grid={grid} />);
    const cols = wrapper.getAllByTestId("cell");
    expect(cols).toMatchSnapshot();
  });
});
