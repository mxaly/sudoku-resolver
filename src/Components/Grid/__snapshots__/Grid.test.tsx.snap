// Vitest Snapshot v1

exports[`Grid > should render the grid 1`] = `
[
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    1
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    2
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    3
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    2
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    3
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    4
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    3
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    4
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    5
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    4
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    5
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    6
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    5
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    6
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    7
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    6
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    7
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    8
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    7
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    8
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    9
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    8
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    9
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    1
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    9
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    1
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    2
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    4
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    5
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    6
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    5
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    6
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    7
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    6
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    7
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    8
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    7
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    8
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    9
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    8
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    9
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    1
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    9
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    1
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    2
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    1
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    2
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    3
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    2
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    3
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    4
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    3
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    4
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    5
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    7
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    8
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    9
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    8
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    9
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    1
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    9
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    1
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    2
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    1
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    2
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    3
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    2
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    3
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    4
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    3
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    4
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    5
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    4
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    5
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    6
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    5
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    6
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    7
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    6
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    7
  </div>,
  <div
    class="_cell_5111c9 _isConstant_5111c9"
    data-testid="cell"
  >
    8
  </div>,
]
`;
